### PR TITLE
Add an absolute pixel coordinates option to grid_sample

### LINKED ISSUE #36107: Allow grid_sample to accept pixel units (absolute coordinates)
_This issue is expanded from #24870 for reference and for additional discussions on implementation details. See that issue for context._

Add an option for the grid/flow input to `grid_sample` to be expressed in pixel units (absolute coordinates). This is much more natural for some applications, including image registration, where the grid-like input is actually a displacement field, and this eliminates the need for the user to convert to normalized [-1,1] coordinates, which is prone to user error, (especially when users don’t know which setting of `align_corners` to target).

This would help simplify the implementation of convolutional networks to do Optical Flow Estimation and Image Registration tasks.

### PR DESCRIPTION
**Pull request by bnehoran**
This adds a new option `pixel_coords` to `nn.functional.grid_sample()`, which, when set to `True`, allows the `grid` passed in to be interpreted in absolute pixel coordinates (unnormalized coordinates).
This is a feature that's been requested by a number of users, and would simplify the implementation of certain architectures for optical flow and image registration tasks.

Closes #36107

When `pixel_coords=False`, which is set as the default, `grid_sample()` behaves exactly as before, with the `grid` being interpreted in normalized coordinates, which means that its values are normalized by the input's dimensions to the range `[-1, 1]`.
This means that in the spatial variant, a grid value of `(-1, -1)` samples from the top-left corner, while a grid value of `(1, 1)` samples from the bottom-right corner of the `input`.

On the other hand, when `pixel_coords=True`, it instead considers the `grid` values as mimicking indices into the input tensor, having the range `[0, size-1]`, but taking possibly non-integral values.
A grid value of `(0, 0)` then samples from the top-left pixel, while a grid value of `(W_in - 1, H_in - 1)` samples from the bottom-right pixel, where `W_in` and `H_in` are the width and height, respectively, of the `input` to be sampled from.

The implementation is rather simple, as all it has to do, really, is just to bypass the `unnormalize()` step in the underlying kernel whenever `pixel_coords` is True. 
The rest of the changes are mostly just updating the function signatures to pass the option down to the kernel.
I've also added tests and updated the docs to reflect the new option.

### CODE REVIEW

**Comment by soumith** [comment]
i'll defer the review to @fmassa

File: aten/src/ATen/native/cpu/GridSamplerKernel.cpp [hunk 1]
```diff
@@ -369,18 +377,31 @@ struct ComputeLocation<scalar_t, GridSamplerPadding::Border, align_corners>
...
   }
 
   inline std::pair<Vec, Vec> apply_get_grad(const Vec &in) const {
-    Vec res, grad_clip;
-    std::tie(res, grad_clip) = clip_coordinates_get_grad(unnormalize(in));
-    return std::make_pair(res, grad_clip & Vec(scaling_factor));
+    Vec res, grad_clip, grad;
+    if (pixel_coords) {
+      res = in;
+      grad = Vec(1);
```
Comment (soumith) [inline aten/src/ATen/native/cpu/GridSamplerKernel.cpp:393]: just checking that this is not `Vec(scaling_factor)`

**Review by soumith — APPROVED** [review]
lgtm, seems fairly mechanical as a PR

**Comment by soumith** [comment]
@gchanan @houseroad what does one have to do to get the Backward Compatibility tests to pass? this PR adds a new option at the end with a default, so it technically should be backward compatible.

File: aten/src/ATen/native/cpu/GridSamplerKernel.cpp (hunk 1 above)
Comment (bnehoran) [reply aten/src/ATen/native/cpu/GridSamplerKernel.cpp:393]: The reason that this is `Vec(1)` here is because when the grid is given in pixel coordinates there is no rescaling (the rescaling happens in `unnormalize()`), so the gradient doesn't need to be multiplied by anything.
That's why it gets multiplied by 1 here instead of by `scaling_factor` (which is the scaling factor by which `unnormalize()` rescales).

Same thing with [line 361](https://github.com/pytorch/pytorch/blob/201b0e7ea171d68b0729bcfa5cbdb7e4b6ee4ea2/aten/src/ATen/native/cpu/GridSamplerKernel.cpp#L361) above.

**Comment by bnehoran** [comment]
@soumith Thanks for the review!

The CI seems to be complaining about the lack of BC for internal functions that aren't meant to be user facing (`aten::grid_sampler_2d`, `aten::grid_sampler_2d_backward`, etc.) and which are therefore not given default values (not for this newly added option nor for any existing ones).
Technically, these internal functions are "exposed" in that users could in theory call them if they made a concerted effort to, but they really shouldn't, and it would be misuse.

Any advice on how to handle this and satisfy the CI's BC check would be appreciated.

File: aten/src/ATen/native/native_functions.yaml [hunk 2]
```diff
@@ -1372,28 +1372,28 @@
 # `interpolation_mode` because it only supports Bilinear interpolation mode.
 # Nor does it take in `align_corners` because it only supports the mode
 # `align_corners = True`.
-- func: grid_sampler(Tensor input, Tensor grid, int interpolation_mode, int padding_mode, bool align_corners) -> Tensor
+- func: grid_sampler(Tensor input, Tensor grid, int interpolation_mode, int padding_mode, bool align_corners, bool pixel_coords) -> Tensor
```
Comment (houseroad) [inline aten/src/ATen/native/native_functions.yaml:1375]: Can we provide a default value for pixel_coords, according to my understanding, if we use pixel_coords, we will keep the old behavior, which means backward compatible (Please correct me if I am wrong).

If we don't provide default value, we will have issues with serialized TorchScript models which contain grid_sampler op.

File: aten/src/ATen/native/native_functions.yaml [hunk 3]
```diff
@@ -1372,28 +1372,28 @@
...
   use_c10_dispatcher: full
 
-- func: grid_sampler_2d(Tensor input, Tensor grid, int interpolation_mode, int padding_mode, bool align_corners) -> Tensor
+- func: grid_sampler_2d(Tensor input, Tensor grid, int interpolation_mode, int padding_mode, bool align_corners, bool pixel_coords) -> Tensor
```
Comment (houseroad) [inline aten/src/ATen/native/native_functions.yaml:1378]: Ditto

File: aten/src/ATen/native/native_functions.yaml [hunk 4]
```diff
@@ -1372,28 +1372,28 @@
...
   use_c10_dispatcher: full
   dispatch:
     CPU: grid_sampler_2d_cpu
     CUDA: grid_sampler_2d_cuda
 
-- func: grid_sampler_2d_backward(Tensor grad_output, Tensor input, Tensor grid, int interpolation_mode, int padding_mode, bool align_corners) -> (Tensor, Tensor)
+- func: grid_sampler_2d_backward(Tensor grad_output, Tensor input, Tensor grid, int interpolation_mode, int padding_mode, bool align_corners, bool pixel_coords) -> (Tensor, Tensor)
```
Comment (houseroad) [inline aten/src/ATen/native/native_functions.yaml:1384]: ditoo

File: aten/src/ATen/native/native_functions.yaml [hunk 5]
```diff
@@ -1372,28 +1372,28 @@
...
   use_c10_dispatcher: full
   dispatch:
     CPU: grid_sampler_2d_backward_cpu
     CUDA: grid_sampler_2d_backward_cuda
 
-- func: grid_sampler_3d(Tensor input, Tensor grid, int interpolation_mode, int padding_mode, bool align_corners) -> Tensor
+- func: grid_sampler_3d(Tensor input, Tensor grid, int interpolation_mode, int padding_mode, bool align_corners, bool pixel_coords) -> Tensor
```
Comment (houseroad) [inline aten/src/ATen/native/native_functions.yaml:1390]: ditto

File: aten/src/ATen/native/native_functions.yaml [hunk 6]
```diff
@@ -1372,28 +1372,28 @@
...
   use_c10_dispatcher: full
   dispatch:
     CPU: grid_sampler_3d_cpu
     CUDA: grid_sampler_3d_cuda
 
-- func: grid_sampler_3d_backward(Tensor grad_output, Tensor input, Tensor grid, int interpolation_mode, int padding_mode, bool align_corners) -> (Tensor, Tensor)
+- func: grid_sampler_3d_backward(Tensor grad_output, Tensor input, Tensor grid, int interpolation_mode, int padding_mode, bool align_corners, bool pixel_coords) -> (Tensor, Tensor)
```
Comment (houseroad) [inline aten/src/ATen/native/native_functions.yaml:1396]: ditto

**Review by houseroad — COMMENTED** [review]
Ideally, we should add default value to make the ops backward compatible.

File: aten/src/ATen/native/native_functions.yaml (hunk 2 above)
Comment (gerardsn) [reply aten/src/ATen/native/native_functions.yaml:1375]: @bnehoran please confirm if the suggested defaults (here and below) are backward compatible. 
```suggestion
- func: grid_sampler(Tensor input, Tensor grid, int interpolation_mode, int padding_mode, bool align_corners, bool pixel_coords=false) -> Tensor
```

File: aten/src/ATen/native/native_functions.yaml (hunk 3 above)
Comment (gerardsn) [reply aten/src/ATen/native/native_functions.yaml:1378]:
```suggestion
- func: grid_sampler_2d(Tensor input, Tensor grid, int interpolation_mode, int padding_mode, bool align_corners, bool pixel_coords=false) -> Tensor
```

File: aten/src/ATen/native/native_functions.yaml (hunk 4 above)
Comment (gerardsn) [reply aten/src/ATen/native/native_functions.yaml:1384]:
```suggestion
- func: grid_sampler_2d_backward(Tensor grad_output, Tensor input, Tensor grid, int interpolation_mode, int padding_mode, bool align_corners, bool pixel_coords=false) -> (Tensor, Tensor)
```

File: aten/src/ATen/native/native_functions.yaml (hunk 5 above)
Comment (gerardsn) [reply aten/src/ATen/native/native_functions.yaml:1390]:
```suggestion
- func: grid_sampler_3d(Tensor input, Tensor grid, int interpolation_mode, int padding_mode, bool align_corners, bool pixel_coords=false) -> Tensor
```

File: aten/src/ATen/native/native_functions.yaml (hunk 6 above)
Comment (gerardsn) [reply aten/src/ATen/native/native_functions.yaml:1396]:
```suggestion
- func: grid_sampler_3d_backward(Tensor grad_output, Tensor input, Tensor grid, int interpolation_mode, int padding_mode, bool align_corners, bool pixel_coords=false) -> (Tensor, Tensor)
```
